docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,811 @@
+  if (err != nil) {
+    return fmt.Errorf(
+      "error parsing extended public key: [%s]",
+      err,
+    )
+  }
+  // For later usage---this is xpub/ypub/zpub/...
+  publicKeyDescriptor = extendedPublicKey[0:4]
+
+  externalChain := extendedKey
+  if (externalChain.Depth() < 4) {
+    // Descend to the external chain path, /0.
+    externalChain, err = extendedKey.Child(0)
+    if (err != nil) {
+      return fmt.Errorf(
+        "error deriving external chain path /0 from extended key: [%s]",
+        err,
+      )
+    }
+  }
+
+  requestedPublicKey, err := externalChain.Child(addressIndex)
+  if (err != nil) {
+    return fmt.Errorf(
+      "error deriving requested address index /0/%s from extended key: [%s]",
+      addressIndex,
+      err,
+    )
+  }
+
+  // Now to decide how we want to serialize the address...
+  var chainParams *chaincfg.Params
+  switch extendedPublicKey[0:4] {
+  case "xpub", "ypub", "zpub":
+    chainParams = &chaincfg.MainNetParams
+  case "tpub", "upub", "vpub":
+    chainParams = &chaincfg.TestNet3Params
+  }
+
+  requestedAddress, err := requestedPublicKey.Address(chainParams)
+
+  var finalAddress btcutil.Address = requestedAddress
+  switch extendedPublicKey[0:4] {
+  case "xpub", "tpub":
+    // Noop, the address is already correct
+  case "ypub", "upub":
+    // p2wpkh-in-p2sh, constructed as per https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#p2wpkh-nested-in-bip16-p2sh .
+    scriptSig = append([]byte{0x00, 0x14}, requestedAddress.Hash160()[:]...)
+    finalAddress = btcutil.NewAddressWitnessScriptHash(
+      btcutil.Hash160(scriptSig),
+      chainParams,
+    )
+  case "zpub", "vpub":
+    // p2wpkh
+    finalAddress, err = btcutil.NewAddressWitnessPubKeyHash(
+      requestedAddress.Hash160()[:],
+      chainParams,
+    )
+  }
+  if (err != nil) {
+    return fmt.Errorf(
+      "failed to derive final address format from extended key: [%s]",
+      err,
+    )
+  }
+
+  return finalAddress.EncodeAddress()
+}
+```
+
+[[sighash-construction]]
+==== Construction of Sighash for Signing
+
+The sighash is constructed as the double-SHA256 of the preimage. The preimage
+is defined in BIP143. The sample function below shows the high level needed
+to construct the sighash, which is the data then signed by the signer group
+to produce a valid transaction (see the section on
+<<final-transaction-construction>> for the rest of the process):
+
+```go
+// import "bytes"
+// import "crypto/sha256"
+// import "encoding/binary"
+// import "github.com/btcsuite/btcutil"
+// import "github.com/btcsuite/btcd/txscript"
+
+// Output is a simplified Bitcoin transaction output that contains a value and a
+// string address.
+type output struct {
+  value uint64
+  addressString string
+}
+
+func doubleSha256(bytes []byte) []byte {
+  hash := sha256.Sum256(bytes)
+  secondHash := sha256.Sum256(hash[:])
+
+  return secondHash[:]
+}
+
+func ConstructSighash(
```

Comment:
Could we first assemble unsigned TX as per `Construction of Final Transaction`, then get its' `sigHash` with btcd's function (instead of constructing it manually here), sign it, and add the signature to the unsigned TX to get the final one?

Unfortunately, I don't remember where in btcd was the function to pull sigHash for unsigned tx :(

Reply:
I couldn't find one… Which is why I did this. This actually flags something I didn't notice that is kind of confusing though. The current approach [recomputes the fee once the tx is fully assembled](https://github.com/keep-network/keep-ecdsa/pull/670/files/29e6d6d266b23e56ac172bb286ca0b1349bfecc4#diff-e19b702a5d320598532e72f46551a840ee1bda9df2e4415f4daf7295441ae354R689), but that shouldn't work given that the sighash includes the output values. I'm thinking I fumbled something when I was testing these against real examples.

Reply:
@beaushinkle @pdyraga I do want to fix this before we merge here. I think the changes will be:
- Compute the fee assuming 175 vBytes (as already [noted for the 5% calculation](https://github.com/keep-network/keep-ecdsa/pull/670/files#diff-e19b702a5d320598532e72f46551a840ee1bda9df2e4415f4daf7295441ae354R237)).
- Update the transaction construction to skip the fee calculation and instead take the fee as a parameter.

The other option is to leave transaction construction as-is and follow @nkuba's approach as a pre step to the sighash by constructing the unsigned transaction, then either finding the secret function to pull the sighash or pulling the fee information from the unsigned transaction. Since this is p2wpkh, I believe the witness script length should be the same no matter what, so the unsigned transaction with a placeholder for signature data should be computable once we have the addresses and desired fees for everyone.

Reply:
https://pkg.go.dev/github.com/btcsuite/btcd/txscript#CalcWitnessSigHash

You beautiful beautiful man @nkuba . Still need to do some fiddling about to get it done with the data we have, but this is going to feel much cleaner.

Reply:
I knew I've seen it somewhere before (and probably used it too)!

Reply:
This is now handled 🎊